File: sorting/leetcode 88. Merge Sorted Array.py
from typing import List

class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        p1 = m-1
        p2 = n-1
        pointer = m+n-1 #指針只到nums1最後面
        while p2>=0: 
            if p1>=0 and nums1[p1]>nums2[p2]: #從nums1最後面的地方開始補數字，nums1的當前指針數比較大就補nums1 反之則是nums2
                nums1[pointer] = nums1[p1]
                p1 -= 1
            else:
                nums1[pointer] = nums2[p2]
                p2 -= 1
            pointer -= 1

        return nums1
'''
本題最佳解
是從兩個陣列的最尾端，也就是兩陣列的最大值開始計算
再併入原本的第一個陣列
從第一個陣列最後面的地方開始補數字
nums1的當前指針數比較大就補nums1 反之則是nums2
'''

nums1 = [1,2,3,0,0,0]
m = 3
nums2 = [2,5,6]
n = 3
solution = Solution()
print(solution.merge(nums1,m,nums2,n))

File: hash table/leetcode 88. Merge Sorted Array.js
class Solution {
    merge(nums1, m, nums2, n) {
        let i = m - 1; // nums1 的最後一個有效元素
        let j = n - 1; // nums2 的最後一個元素
        let k = m + n - 1; // nums1 的最後一個位置

        // 從後往前合併
        while (i >= 0 && j >= 0) {
            if (nums1[i] > nums2[j]) {
                nums1[k] = nums1[i];
                i--;
            } else {
                nums1[k] = nums2[j];
                j--;
            }
            k--;
        }

        // 如果 nums2 中還有剩餘元素，直接複製到 nums1
        while (j >= 0) {
            nums1[k] = nums2[j];
            j--;
            k--;
        }
    }
}

// 測試範例
let nums1 = [1, 2, 3, 0, 0, 0];
let m = 3;
let nums2 = [2, 5, 6];
let n = 3;

let solution = new Solution();
solution.merge(nums1, m, nums2, n);
console.log(nums1); // 輸出: [1, 2, 2, 3, 5, 6]

/*
合併兩個有序數組

重點筆記：
1. **從後往前合併**:
   - 由於 `nums1` 的後半部分是空的，從後往前合併可以避免覆蓋 `nums1` 中的有效元素。

2. **指針的使用**:
   - `i`：指向 `nums1` 的最後一個有效元素。
   - `j`：指向 `nums2` 的最後一個元素。
   - `k`：指向 `nums1` 的最後一個位置。

3. **時間複雜度**:
   - 只需遍歷 `nums1` 和 `nums2` 一次，時間複雜度為 O(m + n)。

4. **空間複雜度**:
   - 只使用了常數級別的額外空間，空間複雜度為 O(1)。

5. **適用場景**:
   - 適合處理大規模數據，因為時間和空間複雜度都很低。
   - 例如：合併兩個有序列表、數據庫中的合併操作等。

範例解析：
- 輸入：nums1 = [1, 2, 3, 0, 0, 0], m = 3, nums2 = [2, 5, 6], n = 3
- 過程：
  - 比較 3 和 6，將 6 放入 nums1[5]，k = 4, j = 2
  - 比較 3 和 5，將 5 放入 nums1[4]，k = 3, j = 1
  - 比較 3 和 2，將 3 放入 nums1[3]，k = 2, i = 2
  - 比較 2 和 2，將 2 放入 nums1[2]，k = 1, j = 0
  - 比較 2 和 1，將 2 放入 nums1[1]，k = 0, i = 1
  - 將剩餘的 1 放入 nums1[0]
- 輸出：nums1 = [1, 2, 2, 3, 5, 6]
*/